src/k2/dto/SKVRecord.h

```diff
@@ -75,24 +79,25 @@ class SKVRecord {
         }
 
         FieldType ft = TToFieldType<T>();
-        if (fieldCursor >= schema->fields.size() || ft != schema->fields[fieldCursor].type) {
-            throw TypeMismatchException("Schema not followed in record serialization");
+        if (serializerFieldCursor >= schema->fields.size() || ft != schema->fields[serializerFieldCursor].type) {
+            throw TypeMismatchException(fmt::format("Schema not followed in record serialization for index {}",
+                                        serializerFieldCursor));
         }
 
         for (size_t i = 0; i < schema->partitionKeyFields.size(); ++i) {
-            if (schema->partitionKeyFields[i] == fieldCursor) {
+            if (schema->partitionKeyFields[i] == serializerFieldCursor) {
                 partitionKeys[i] = FieldToKeyString<T>(field);
             }
         }
 
         for (size_t i = 0; i < schema->rangeKeyFields.size(); ++i) {
-            if (schema->rangeKeyFields[i] == fieldCursor) {
+            if (schema->rangeKeyFields[i] == serializerFieldCursor) {
                 rangeKeys[i] = FieldToKeyString<T>(field);
             }
         }
 
         storage.fieldData.write(field);
-        ++fieldCursor;
+        ++serializerFieldCursor;
```

Comment:
In this case we have advanced the underlying payload cursor, so deserialzeFieldCursor should be incremented as well. Otherwise the user might expect to be able to deserialize the value just serialized, but that isn't possible without calling seekField first. In general, the deserializeFieldCursor should always match the last field serialized regardless of the payload cursor and the serializeFieldCursor should match the payload cursor. So calling serializeNext may require calling seekField and will necessarily require incrementing the deserializeFieldCursor.

Reply:
We should also have a unit tests for serializing and deserializing the same SKVRecord which will test the logic in the above comment.